homeassistant/components/xiaomi_miio/config_flow.py

```diff
@@ -101,76 +202,204 @@ async def async_step_zeroconf(self, discovery_info):
+        if self.model is None:
+            errors["base"] = "cannot_connect"
+            return await self.async_step_model(errors=errors)
+
+        if self.mac is None and device_info is not None:
+            self.mac = format_mac(device_info.mac_address)
+
+        unique_id = self.mac
+        existing_entry = await self.async_set_unique_id(
+            unique_id, raise_on_progress=False
+        )
+        if existing_entry:
+            data = existing_entry.data.copy()
+            data[CONF_HOST] = self.host
+            data[CONF_TOKEN] = self.token
+            if (
+                self.cloud_username is not None
+                and self.cloud_password is not None
+                and self.cloud_country is not None
+            ):
+                data[CONF_CLOUD_USERNAME] = self.cloud_username
+                data[CONF_CLOUD_PASSWORD] = self.cloud_password
+                data[CONF_CLOUD_COUNTRY] = self.cloud_country
+            self.hass.config_entries.async_update_entry(existing_entry, data=data)
+            await self.hass.config_entries.async_reload(existing_entry.entry_id)
+            return self.async_abort(reason="reauth_successful")
+
+        # Setup Gateways
+        for gateway_model in MODELS_GATEWAY:
+            if self.model.startswith(gateway_model):
+                if self.name is None:
+                    self.name = DEFAULT_GATEWAY_NAME
+                return self.async_create_entry(
+                    title=self.name,
+                    data={
+                        CONF_FLOW_TYPE: CONF_GATEWAY,
+                        CONF_HOST: self.host,
+                        CONF_TOKEN: self.token,
+                        CONF_MODEL: self.model,
+                        CONF_MAC: self.mac,
+                        CONF_CLOUD_USERNAME: self.cloud_username,
+                        CONF_CLOUD_PASSWORD: self.cloud_password,
+                        CONF_CLOUD_COUNTRY: self.cloud_country,
+                    },
+                )
+
+        # Setup all other Miio Devices
+        if self.name is None:
+            self.name = self.model
+
+        for device_model in MODELS_ALL_DEVICES:
+            if self.model.startswith(device_model):
+                return self.async_create_entry(
+                    title=self.name,
+                    data={
+                        CONF_FLOW_TYPE: CONF_DEVICE,
+                        CONF_HOST: self.host,
+                        CONF_TOKEN: self.token,
+                        CONF_MODEL: self.model,
+                        CONF_MAC: self.mac,
+                        CONF_CLOUD_USERNAME: self.cloud_username,
+                        CONF_CLOUD_PASSWORD: self.cloud_password,
+                        CONF_CLOUD_COUNTRY: self.cloud_country,
+                    },
+                )
+
+        errors["base"] = "unknown_device"
+        return await self.async_step_model(errors=errors)
+
+    async def async_step_model(self, user_input=None, errors=None):
```

Comment:
We could merge this step with the connect step. Then we wouldn't need to pass errors around.

Reply:
Done, I merged the two